### ISSUE TITLE
Incompatible with pytorch's register_buffer() methods

### ISSUE THREAD

**Issue by YuCao16** [microsoft/pyright#4213]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
[`register_buffer()`](https://pytorch.org/docs/stable/generated/torch.nn.Module.html?highlight=register_buffer#torch.nn.Module.register_buffer) adds a buffer to the module, buffers can be accessed as attributes using given names. However, those attributes will trigger Error: `__getitem__` method not defined on type Module.

**To Reproduce**
```python
from typing import Dict

import torch
from torch import nn


def foo() -> Dict[str, torch.Tensor]:
    alpha_t = torch.tensor([1.0, 1.5])
    return {"alpha_t": alpha_t}


class Test(nn.Module):
    def __init__(self) -> None:
        super(Test, self).__init__()

        for k, v in foo().items():
            self.register_buffer(k, v)

    def test(self) -> torch.Tensor:
        print(self.alpha_t)
        return self.alpha_t[0]

```
and the Output:
```
tensor([1.0000, 1.5000])                                                                                                                               
tensor(1.) 
```

**Expected behavior**
There should be no problem with these lines of code, so it should not prompt an Error.

**Screenshots or Code**
As above.

If your code relies on symbols that are imported from a third-party library, include the associated import statements and specify which versions of those libraries you have installed.

**VS Code extension or command-line**
pyright command-line with version 1.1.280.

**Additional context**
However, `__getitem__` comes from `self.alpha_t[0]`, if change to `self.alpha_t`, Error message:
```
Expression of type "Tensor | Module" cannot b
e assigned to return type "Tensor"
    Type "Tensor | Module" cannot be assigned to type "Tensor"
      "Module" is incompatible with "Tensor" (reportGeneralTypeIssues)
```
I know I can add `#pyright: ignore` to disable the error message on this line, but is there an update or type annotation can be added to avoid this type of error?

**Comment by erictraut** [comment]
Pyright is doing the right thing here. The `nn.Module` class defines a `__getattr__` method that returns `Tensor | Module`.

```python
    def __getattr__(self, name: str) -> Union[Tensor, 'Module']:
```

If you know that the `alpha_t` attribute is a `Tensor` rather than a `Module`, you can add a runtime assert. Pyright will use this to narrow the type to `Tensor`, which eliminates the error.

```python
    def test(self) -> torch.Tensor:
        print(self.alpha_t)
        assert isinstance(self.alpha_t, torch.Tensor)
        return self.alpha_t[0]
```

**Comment by YuCao16** [comment]
> Pyright is doing the right thing here. The `nn.Module` class defines a `__getattr__` method that returns `Tensor | Module`.
> 
> ```python
>     def __getattr__(self, name: str) -> Union[Tensor, 'Module']:
> ```
> 
> If you know that the `alpha_t` attribute is a `Tensor` rather than a `Module`, you can add a runtime assert. Pyright will use this to narrow the type to `Tensor`, which eliminates the error.
> 
> ```python
>     def test(self) -> torch.Tensor:
>         print(self.alpha_t)
>         assert isinstance(self.alpha_t, torch.Tensor)
>         return self.alpha_t[0]
> ```

Thanks for your explanation!

**Comment by chriskeraly-cruise** [comment]
I understand that is correct, but requiring asserts every time we want to use a buffer variable is cumbersome and pollutes the code more than anything else.

Seems like pyre does some fancy stuff to avoid this. https://pyre-check.org/docs/features/

**Comment by hmc-cs-mdrissi** [comment]
I’d recommend having it be Any as return type instead of getattr. A union return type in general is awkward to work with. If one type, A, is much more dominant then others sometimes typeshed will do A | Any, but if multiple types are common then you are stuck between be safe and have a bunch of assertions with union return type of be a bit laxer when getattr is hit and have an Any return type.
